File: frontend/src/api/patternDrill.ts
import { api } from "./client";
import type { DrillCard } from "../types";

export interface DrillDeckResponse {
  items: DrillCard[];
  total: number;
}

export function fetchPatternDeck(topic?: string, difficulty?: string, limit = 50) {
  const params = new URLSearchParams();
  if (topic) params.set("topic", topic);
  if (difficulty) params.set("difficulty", difficulty);
  params.set("limit", String(limit));
  return api.get<DrillDeckResponse>(`/pattern-drill/deck?${params.toString()}`);
}

export interface DrillReviewResponse {
  question_id: number;
  repetitions: number;
  interval: number;
  next_review: string | null;
  last_reviewed: string | null;
}

export function reviewDrill(questionId: number, quality: number, notes?: string) {
  return api.post<DrillReviewResponse>(`/pattern-drill/${questionId}/review`, {
    quality,
    notes: notes || undefined,
  });
}

export interface DrillAskResponse {
  id: number;
  question_id: number;
  user_question: string;
  answer: string;
  created_at: string;
}

export function askDrill(questionId: number, question: string, stepKind?: string) {
  return api.post<DrillAskResponse>(`/pattern-drill/${questionId}/ask`, {
    question,
    step_kind: stepKind || undefined,
  });
}

export function fetchPatternAnalysis(questionId: number) {
  return api.get<Record<string, unknown>>(`/pattern-drill/${questionId}/analysis`);
}

export function generatePatternAnalysis(questionId: number) {
  return api.post<DrillCard>(`/pattern-drill/${questionId}/generate`);
}

export function regeneratePatternAnalysis(questionId: number) {
  return api.post<DrillCard>(`/pattern-drill/${questionId}/regenerate`);
}

export function regenerateAllPatternAnalyses() {
  return api.post<{ status: string; question_count: number }>("/pattern-drill/regenerate-all");
}

File: backend/schemas/pattern_drill.py
from datetime import datetime

from pydantic import BaseModel

from backend.models.question import Difficulty
from backend.schemas.question import PatternAnalysis


class DrillCard(BaseModel):
    id: int
    number: int | None
    title: str
    difficulty: Difficulty
    topics: list[str]
    subtopics: list[str]
    pattern_analysis: PatternAnalysis | None = None
    completed: bool = False


class DrillDeckResponse(BaseModel):
    items: list[DrillCard]
    total: int


class DrillReviewRequest(BaseModel):
    quality: int
    notes: str | None = None


class DrillReviewResponse(BaseModel):
    question_id: int
    repetitions: int
    interval: int
    next_review: datetime | None = None
    last_reviewed: datetime | None = None


class DrillAskRequest(BaseModel):
    question: str
    step_kind: str | None = None


class DrillAskResponse(BaseModel):
    id: int
    question_id: int
    user_question: str
    answer: str
    created_at: datetime

    model_config = {"from_attributes": True}
